### PR TITLE
[CUBRID] Display Partition Tables

### LINKED ISSUE #34168: [CUBRID] Display Partition Tables
**Description**
Currently, **Partition Table** is display as normal **Table**. 
**Partition table** should not be shown in the navigation list and should be included in the normal **Table** instead.

**Solution**
Make a folder **Partitions** under table and display all partitions which is related to each table.

### PR DESCRIPTION
**Pull request by rathana-pvs**
- Make a folder Partitions under table and display all partitions which is related to each table.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package org.jkiss.dbeaver.ext.cubrid.model;
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java:17]: Please add copyright here. You can take it from other files with code.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            JDBCResultSet dbResult) {
+    super(table.getContainer(), name, type,dbResult);
+    this.parentTable = table;
+    
+   
+    expression = JDBCUtils.safeGetString(dbResult, "partition_expr");
+    expression_values = (Integer[]) JDBCUtils.safeGetObject(dbResult, "partition_values");
+    comment = JDBCUtils.safeGetString(dbResult, "comment");
+    this.setDescription(comment);
+  }
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]: Formatting is broken here
```suggestion
    public CubridPartition(
        @NotNull CubridTable table,
        @NotNull String name,
        @NotNull String type,
        @NotNull JDBCResultSet dbResult
    ) {
        super(table.getContainer(), name, type,dbResult);
        this.parentTable = table;
        expression = JDBCUtils.safeGetString(dbResult, "partition_expr");
        expression_values = (Integer[]) JDBCUtils.safeGetObject(dbResult, "partition_values");
        comment = JDBCUtils.safeGetString(dbResult, "comment");
        this.setDescription(comment);
    }
```

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java [hunk 3]
```diff
@@ -115,4 +127,28 @@ public Object[] getPossibleValues(CubridTable object)
...
+        protected CubridPartition fetchObject(
+            JDBCSession session, CubridTable table, JDBCResultSet dbResult)
+            throws SQLException, DBException {
+            String partition_class_name = JDBCUtils.safeGetString(dbResult, "partition_class_name");
+            String type = JDBCUtils.safeGetString(dbResult, "partition_type");
+            
+            return new CubridPartition(table, partition_class_name, type, dbResult);
+        }
+            
+        }
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java]: Please fix formatting
```suggestion
    class PartitionCache extends JDBCObjectCache<CubridTable, CubridPartition>{

        @Override
        protected JDBCStatement prepareObjectsStatement(@NotNull JDBCSession session, @NotNull CubridTable table) throws SQLException {
            String sql = "select * from db_partition where class_name = ? and owner_name = ?";
            final JDBCPreparedStatement dbStat = session.prepareStatement(sql);
            dbStat.setString(1, table.getName());
            dbStat.setString(2, table.getSchema().getName());
            return dbStat;
        }
    
        @Override
        protected CubridPartition fetchObject(
            @NotNull JDBCSession session,
            @NotNull CubridTable table,
            @NotNull JDBCResultSet dbResult
        ) throws SQLException, DBException {
            String partition_class_name = JDBCUtils.safeGetString(dbResult, "partition_class_name");
            String type = JDBCUtils.safeGetString(dbResult, "partition_type");
            
            return new CubridPartition(table, partition_class_name, type, dbResult);
        }
            
    }

```

**Review by E1izabeth — CHANGES_REQUESTED** [review]

**Comment by E1izabeth** [comment]
@rathana-pvs 
Thank you for contribution. I reviewed your PR and some changes should be done.

File: plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml [hunk 4]
```diff
@@ -57,6 +57,12 @@
                                         </items>
                                     </items>
                                 </folder>
+                                <folder type="org.jkiss.dbeaver.ext.cubrid.model.CubridPartition" label="%tree.partitions.node.name" icon="#partitions" description="%tree.partitions.node.tip">
+                                    <items label="%tree.partitions.node.name" path="partition" property="partitions" icon="#partition">
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:61]: Like in the Postgres bundle - it has to be items label="%tree.partition.node.name"
And also register tree.partition.node.name in the plugins/org.jkiss.dbeaver.ext.cubrid/OSGI-INF/l10n/bundle.properties
It all for navigator tree and properties items/commands

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        return this.parentTable;
+    }
+    @Override 
+    public boolean isSubPartition(){
+        return false;
+    }
+    
+    @Override 
+    public DBSTablePartition getPartitionParent(){
+        return null;
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java:57]: just to check - no subpartitions in Cubrid?

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        @NotNull String name,
+        @NotNull String type,
+        @NotNull JDBCResultSet dbResult
+    ) {
+        super(table.getContainer(), name, type,dbResult);
+        this.parentTable = table;
+        expression = JDBCUtils.safeGetString(dbResult, "partition_expr");
+        expression_values = (Integer[]) JDBCUtils.safeGetObject(dbResult, "partition_values");
+        comment = JDBCUtils.safeGetString(dbResult, "comment");
+        this.setDescription(comment);
```
Comment (LonwoLonwo) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]: Why do you want to set description if you have this method?
`public String getDescription()`

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java (hunk 5 above)
Comment (rathana-pvs) [reply plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java:57]: Currently, Cubrid doesn't support sub-partition.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java (hunk 6 above)
Comment (rathana-pvs) [reply plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]: Thank you for checking it.
It is my mistake that forgot to remove it.

File: plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml (hunk 4 above)
Comment (rathana-pvs) [reply plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:61]: I have followed the same **label** and register **bundle.properties** as Postgres. 

Do you mean changing from "%tree.**partitions**.node.name" to "%tree.**partition**.node.name"?

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import org.jkiss.dbeaver.model.meta.Property;
+import org.jkiss.dbeaver.model.struct.rdb.DBSTable;
+import org.jkiss.dbeaver.model.struct.rdb.DBSTablePartition;
+
+public class CubridPartition extends CubridTable implements DBSTablePartition{
+
+    private CubridTable parentTable;
+    private String expression;
+    private Integer[] expression_values;
+    private String comment;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]: Please mark these `final` and rename `expression_values` to `expressionValues`

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+import org.jkiss.code.NotNull;
+import org.jkiss.code.Nullable;
+import org.jkiss.dbeaver.model.exec.jdbc.JDBCResultSet;
+import org.jkiss.dbeaver.model.impl.jdbc.JDBCUtils;
+import org.jkiss.dbeaver.model.meta.Property;
+import org.jkiss.dbeaver.model.struct.rdb.DBSTable;
+import org.jkiss.dbeaver.model.struct.rdb.DBSTablePartition;
+
+public class CubridPartition extends CubridTable implements DBSTablePartition{
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]:
```suggestion
public class CubridPartition extends CubridTable implements DBSTablePartition {
```

File: plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml (hunk 4 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:61]: Correct.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java (hunk 8 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridPartition.java]: Additionally, please specify nullability annotations for methods from base classes and fix code style.

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java [hunk 9]
```diff
@@ -64,6 +70,12 @@ public Collection<? extends GenericTableIndex> getIndexes(DBRProgressMonitor mon
     {
         return this.getDataSource().getObjectContainer().getCubridIndexCache().getObjects(monitor, getContainer(), this);
     }
+    
+    public Collection<CubridPartition> getPartitions(DBRProgressMonitor monitor)
+            throws DBException
+    {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java]:
```suggestion
    @NotNull
    public Collection<CubridPartition> getPartitions(@NotNull DBRProgressMonitor monitor) throws DBException {

```

File: plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java [hunk 10]
```diff
@@ -115,4 +127,30 @@ public Object[] getPossibleValues(CubridTable object)
             return object.getDataSource().getSchemas().toArray();
         }
     }
+    
+    class PartitionCache extends JDBCObjectCache<CubridTable, CubridPartition>{
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java]:
```suggestion
    static class PartitionCache extends JDBCObjectCache<CubridTable, CubridPartition> {

```

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by uslss** [comment]
error is shown on attempt to open partition tab
image

Test table
```sql
CREATE TABLE participant2 (
    host_year INT,
    nation CHAR(3),
    gold INT,
    silver INT,
    bronze INT
)
PARTITION BY RANGE (host_year) (
    PARTITION before_2000 VALUES LESS THAN (2000),
    PARTITION before_2008 VALUES LESS THAN (2008)
);
```

**Comment by rathana-pvs** [comment]
> error is shown on attempt to open partition tab image
> 
> Test table
> 
> ```sql
> CREATE TABLE participant2 (
>     host_year INT,
>     nation CHAR(3),
>     gold INT,
>     silver INT,
>     bronze INT
> )
> PARTITION BY RANGE (host_year) (
>     PARTITION before_2000 VALUES LESS THAN (2000),
>     PARTITION before_2008 VALUES LESS THAN (2008)
> );
> ```

Thank you for checking!
I found out, It happens on server 11.0. Because it doesn't support multi schema. 
I will fix it soon.

**Review by E1izabeth — APPROVED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**